JavaNetClientBuilder fails with NoContent response

Calling delete on a service that returns NoContent results in an exception when using JavaNetClientBuilder.
The call is 
```
httpClient.status(Request[IO](Method.DELETE, uri))
```
The exception I get is 
```
Caused by: java.io.FileNotFoundException: http://localhost:8080/history/98613681-91f2-4f42-ac23-655de09178b3
        ...
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1490)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1474)
	at org.http4s.client.JavaNetClientBuilder.$anonfun$open$21(JavaNetClientBuilder.scala:120)
       ...
Caused by: java.io.FileNotFoundException: http://localhost:8080/history/98613681-91f2-4f42-ac23-655de09178b3
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1872)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1474)
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
	at org.http4s.client.JavaNetClientBuilder.$anonfun$fetchResponse$2(JavaNetClientBuilder.scala:128)
```
Note that the call succeeds using BlazeClientBuilder

--- a/client/src/main/scala/org/http4s/client/JavaNetClient.scala
+++ b/client/src/main/scala/org/http4s/client/JavaNetClient.scala
@@ -4,6 +4,8 @@ package client
 import cats.data.Kleisli
 import cats.effect.{Effect, Sync}
 import cats.implicits._
+import fs2.Stream
+import java.io.IOException
 import java.net.{HttpURLConnection, Proxy, URL}
 import javax.net.ssl.{HostnameVerifier, HttpsURLConnection, SSLSocketFactory}
 import org.http4s.internal.{blocking, readInputStream, writeOutputStream}
@@ -101,7 +103,12 @@ sealed abstract class JavaNetClient private (
       _ <- F.delay(conn.setInstanceFollowRedirects(false))
       _ <- F.delay(conn.setDoInput(true))
       resp <- blocking(fetchResponse(req, conn), blockingExecutionContext)
-    } yield DisposableResponse(resp, F.delay(conn.getInputStream.close()))
+    } yield {
+      val dispose = F.delay(conn.getInputStream().close()).recoverWith {
+        case _: IOException => F.delay(Option(conn.getErrorStream()).foreach(_.close()))
+      }
+      DisposableResponse(resp, dispose)
+    }
   }
 
   private def fetchResponse[F[_]](req: Request[F], conn: HttpURLConnection)(implicit F: Effect[F]) =
@@ -116,7 +123,7 @@ sealed abstract class JavaNetClient private (
             .flatMap { case (k, vs) => vs.asScala.map(Header(k, _)) }
             .toList
         ))
-      body = readInputStream(F.delay(conn.getInputStream), 4096, blockingExecutionContext)
+      body = readBody(conn)
     } yield Response(status = status, headers = headers, body = body)
 
   private def timeoutMillis(d: Duration): Int = d match {
@@ -151,6 +158,18 @@ sealed abstract class JavaNetClient private (
         case _ =>
           F.delay(conn.setDoOutput(false))
       }
+
+  private def readBody[F[_]](conn: HttpURLConnection)(implicit F: Effect[F]) = {
+    def inputStream =
+      F.delay(Option(conn.getInputStream)).recoverWith {
+        case _: IOException if conn.getResponseCode > 0 =>
+          F.delay(Option(conn.getErrorStream))
+      }
+    Stream.eval(inputStream).flatMap {
+      case Some(in) => readInputStream(F.pure(in), 4096, blockingExecutionContext, false)
+      case None => Stream.empty
+    }
+  }
 
   private def configureSsl[F[_]](conn: HttpURLConnection)(implicit F: Sync[F]) =
     conn match {

--- a/client/src/test/scala/org/http4s/client/ClientRouteTestBattery.scala
+++ b/client/src/test/scala/org/http4s/client/ClientRouteTestBattery.scala
@@ -42,7 +42,7 @@ abstract class ClientRouteTestBattery(name: String, client: Client[IO])
         val port = address.getPort
         val req = Request[IO](uri = Uri.fromString(s"http://$name:$port$path").yolo)
         client
-          .fetch(req)(resp => IO(checkResponse(resp, expected)))
+          .fetch(req)(resp => checkResponse(resp, expected))
           .unsafeRunTimed(timeout)
           .get
       }
@@ -103,16 +103,16 @@ abstract class ClientRouteTestBattery(name: String, client: Client[IO])
       }
     )
 
-  private def checkResponse(rec: Response[IO], expected: Response[IO]) = {
+  private def checkResponse(rec: Response[IO], expected: Response[IO]): IO[Boolean] = {
     val hs = rec.headers.toSeq
-
-    rec.status must be_==(expected.status)
-
-    collectBody(rec.body) must be_==(collectBody(expected.body))
-
-    expected.headers.foreach(h => h must beOneOf(hs: _*))
-
-    rec.httpVersion must be_==(expected.httpVersion)
+    for {
+      _ <- IO(rec.status must be_==(expected.status))
+      body <- rec.body.compile.to[Array]
+      expBody <- expected.body.compile.to[Array]
+      _ <- IO(body must_== expBody)
+      _ <- IO(expected.headers.foreach(h => h must beOneOf(hs: _*)))
+      _ <- IO(rec.httpVersion must be_==(expected.httpVersion))
+    } yield true
   }
 
   private def renderResponse(srv: HttpServletResponse, resp: Response[IO]): Unit = {
@@ -126,7 +126,4 @@ abstract class ClientRouteTestBattery(name: String, client: Client[IO])
       .drain
       .unsafeRunSync()
   }
-
-  private def collectBody(body: EntityBody[IO]): Array[Byte] =
-    body.compile.toVector.unsafeRunSync().toArray
 }

--- a/client/src/test/scala/org/http4s/client/testroutes/GetRoutes.scala
+++ b/client/src/test/scala/org/http4s/client/testroutes/GetRoutes.scala
@@ -12,6 +12,9 @@ object GetRoutes {
   val SimplePath = "/simple"
   val ChunkedPath = "/chunked"
   val DelayedPath = "/delayed"
+  val NoContentPath = "/no-content"
+  val NotFoundPath = "/not-found"
+  val EmptyNotFoundPath = "/empty-not-found"
 
   def getPaths(implicit ec: ExecutionContext): Map[String, Response[IO]] =
     Map(
@@ -19,6 +22,9 @@ object GetRoutes {
       ChunkedPath -> Response[IO](Ok).withBody(
         Stream.emits("chunk".toSeq.map(_.toString)).covary[IO]),
       DelayedPath ->
-        IO.sleep(1.seconds) *> Response[IO](Ok).withBody("delayed path")
+        IO.sleep(1.seconds) *> Response[IO](Ok).withBody("delayed path"),
+      NoContentPath -> Response[IO](NoContent).pure[IO],
+      NotFoundPath -> Response[IO](NotFound).withBody("not found"),
+      EmptyNotFoundPath -> Response[IO](NotFound).pure[IO]
     ).mapValues(_.unsafeRunSync())
 }

